File: clustering_Football.R
#-----------------------------Step 1:Importing Library-------------------------------------------

  library(ggplot2)
  library(ggdendro)

#-----------------------------Step 2:Read CSV file-----------------------------------------------

  setwd("W:\\R code\\Clustering")
  FootballPlayerRecords <- read.csv("Song_Football.csv")
  View(FootballPlayerRecords)

#----------------------------Step 3: Preparing Data quality Report-------------------------------
  str(FootballPlayerRecords)
  summary(FootballPlayerRecords)
  
  DataTypes <-1
  Maxs<-1
  Mins<-1
  Average<-1
  SD<-1
  U<-1
  
  variable<-names(FootballPlayerRecords)
  
  for(i in 1:length(variable))
  {
    x<-variable[i]
    y<-FootballPlayerRecords[,x] #all the values present in a given column
    
    DataTypes[i] <-class(y)
    Maxs[i]<-ifelse((class(y)=="integer" | class(y)=="numeric"),max(y,na.rm=TRUE),0)
    Mins[i]<-ifelse((class(y)=="integer" | class(y)=="numeric"),min(y,na.rm=TRUE),0)
    Average[i]<-as.integer(ifelse((class(y)=="integer" | class(y)=="numeric"),mean(y,na.rm=TRUE),0))
    SD[i]<-ifelse((class(y)=="integer" | class(y)=="numeric"),sd(y,na.rm=TRUE),0)
    U[i]<-ifelse(class(y)=="factor" ,unique(y),0)
    
  }
  
  missing_count<-colSums(is.na(FootballPlayerRecords))
  
  final_summary<-data.frame(variable=names(FootballPlayerRecords),data_type=DataTypes,Min=Mins,Max=Maxs,
                            average=Average,SD=SD,unique=U,missing=missing_count)
  
  
  rownames(final_summary) <- NULL
  write.csv(final_summary,file="Data_report.csv")
  
  
#----------------------------Step 4: Missing value treatment--------------------------------------
   # There is no missing, no actions required

#----------------------------Step 5: outlier  treatment--------------------------------------
    # outliers doesn't affect clustering there is no need to remove or impute the oulier
    # If there are outlier they will automatically assign to a single cluster
  
#----------------------------Step 6: Check information about insured player Song -----------------
  
song <- FootballPlayerRecords[which(FootballPlayerRecords$First_Name=='Song'),] 
View(song)


#----------------------------Step 7: Data preparation for clutering------------------------------

  # keeping only the numeric values
  # for clustering we don't need 1st column and use col 2:8
FootballPlayerRecords_Clustering <- FootballPlayerRecords[,2:8]

# scaling ( ordering is same)
FootballPlayerRecords_Clustering <- scale(FootballPlayerRecords_Clustering,
                                          center=TRUE,scale=TRUE)

# weighting tackles by 3
FootballPlayerRecords_Clustering[,1] <- FootballPlayerRecords_Clustering[,1]*3


#------------------------Step 8: find the optimal number of clusters( K value) using elbow method----

#Create a screeplot-plot of cluster's tot.withinss wrt number of clusters

  wss<-1:15
  number<-1:15
  
  for (i in 1:15)
    
  {
    wss[i]<-kmeans(FootballPlayerRecords_Clustering,i)$tot.withinss
  }

  
  data<-data.frame(wss,number)
  p<-ggplot(data,aes(x=number,y=wss),color="red")
  p+geom_point()
 # According to Scatterplot , optimal k value =8

#------------------------Step 9: Building cluster and assigning to players records --------------
 
  set.seed(8)
  Clusters_result <- kmeans(FootballPlayerRecords_Clustering ,8)  
  
  FootballPlayerRecords$Cluster_no <- Clusters_result$cluster
  
#------------------------Step 10: Finding players similar to Song -------------------------------
  
  Song_cluster <- FootballPlayerRecords[which(FootballPlayerRecords$First_Name=='Song'),]
  
  Players_similar_to_song <-FootballPlayerRecords[which(FootballPlayerRecords$Cluster_no==
                                                          Song_cluster$Cluster_no),] 
  
  View(Players_similar_to_song)
  

  
#------------------------------Different Approch:Using Hierarchical cluster -------------------
  
  d <- dist(FootballPlayerRecords_Clustering, method = "euclidean")
  
  HCluster_result <- hclust(d,method ="average" )
  plot(HCluster_result,
       main = 'Dendrogram',
       xlab = 'Player',
       ylab = 'Euclidean distances') # display dendogram
  
  ggdendrogram(HCluster_result, 
               rotate = FALSE, 
               theme_dendro = TRUE, 
               color = "tomato")
  
  
  hc_fit = cutree(HCluster_result, 7)
  
  FootballPlayerRecords$Cluster_no <- hc_fit
  
  
  Song_cluster <- FootballPlayerRecords[which(FootballPlayerRecords$First_Name=='Song'),]
  
  Players_similar_to_song <-FootballPlayerRecords[which(FootballPlayerRecords$Cluster_no==
                                                          Song_cluster$Cluster_no),] 
  
  View(Players_similar_to_song)
  
  
  

File: clustering_Football.py
#----------------- Step 1:  Importing required packages for this problem ------------------------------------- 
   # data analysis and wrangling
    import pandas as pd
    import numpy as np
    import random as rn
    from sklearn.preprocessing import StandardScaler
    from sklearn.decomposition import PCA
    
   # K-mean library
    from sklearn.cluster import KMeans
    import scipy.cluster.hierarchy as sch
    from sklearn.cluster import AgglomerativeClustering
   
   # visualization
    import seaborn as sns
    import matplotlib.pyplot as plt

#-----------------------------Step 2:Read CSV file--------------------------------------
      
    # loading train and test sets with pandas 
    FootballPlayerRecords = pd.read_csv('Song_Football.csv',encoding='latin-1')
   
    # Print the columns of dataframe
    print(FootballPlayerRecords.columns.values)
    
    # Returns first n rows
    FootballPlayerRecords.head(10)
    
    
    # Retrive data type of object and no. of non-null object
    FootballPlayerRecords.info()
    
    # Retrive details of integer and float data type 
    FootballPlayerRecords.describe()
    
    # To get  details of the categorical types
    FootballPlayerRecords.describe(include=['O'])

   

 #----------------------------Step 3: Preparing Data quality Report-----------------------

  # To get count of no. of NULL for each data type columns = full.columns.values
    columns = FootballPlayerRecords.columns.values
    data_types = pd.DataFrame(FootballPlayerRecords.dtypes, columns=['data types'])
    
    missing_data_counts = pd.DataFrame(FootballPlayerRecords.isnull().sum(),
                            columns=['Missing Values'])
    
    present_data_counts = pd.DataFrame(FootballPlayerRecords.count(), columns=['Present Values'])
    
    UniqueValues = pd.DataFrame(FootballPlayerRecords.nunique(), columns=['Unique Values'])
    
    MinimumValues = pd.DataFrame(columns=['Minimum Values'])
    for c in list(columns):
       if (FootballPlayerRecords[c].dtypes == 'float64' ) | (FootballPlayerRecords[c].dtypes == 'int64'):
            MinimumValues.loc[c]=FootballPlayerRecords[c].min()
       else:
            MinimumValues.loc[c]=0
 
    MaximumValues = pd.DataFrame(columns=['Maximum Values'])
    for c in list(columns):
       if (FootballPlayerRecords[c].dtypes == 'float64' ) |(FootballPlayerRecords[c].dtypes == 'int64'):
            MaximumValues.loc[c]=FootballPlayerRecords[c].max()
       else:
            MaximumValues.loc[c]=0
    
    data_quality_report=data_types.join(missing_data_counts).join(present_data_counts).join(UniqueValues).join(MinimumValues).join(MaximumValues)
    data_quality_report.to_csv('Data_report.csv', index=True) 
#----------------------------Step 4: Missing value treatment--------------------------------
   # There is no missing, no actions required

#----------------------------Step 5: outlier  treatment-------------------------------------
    # outliers doesn't affect clustering there is no need to remove or impute the oulier
    # If there are outlier they will automatically assign to a single cluster
  
#----------------------------Step 6: Check information about insured player Song -----------------
  
    song = FootballPlayerRecords.loc[FootballPlayerRecords['First_Name']=='Song',] 
    song


#----------------------------Step 7: Data preparation for clutering------------------------------

  # keeping only the numeric values
  # for clustering we don't need 1st column and use col 2:8
  FootballPlayerRecords_Clustering = FootballPlayerRecords.iloc[:,1:8]  
  
  #-----Scaling
  sc = StandardScaler()
  FootballPlayerRecords_Clustering = sc.fit_transform(FootballPlayerRecords_Clustering)
  # weighting tackles by 3
  FootballPlayerRecords_Clustering[:,0] = FootballPlayerRecords_Clustering[:,0] *3


#--------------Step 8: find the optimal number of clusters( K value) using elbow method----
    wcss=[]
   
    for i in range(1, 15):
        kmeans = KMeans(n_clusters = i, init = 'k-means++', random_state = 42)
        kmeans.fit(FootballPlayerRecords_Clustering)
        wcss.append(kmeans.inertia_)
        
    plt.plot(range(1, 15), wcss)
    plt.title('The Elbow Method')
    plt.xlabel('Number of clusters')
    plt.ylabel('WCSS')
    plt.show()
   # According to Scatterplot , optimal k value =8

#-------------------Step 9: Building cluster and assigning to players records --------------
 
  Clusters_result = KMeans(n_clusters = 8, init = 'k-means++', random_state = 42) 
  Clusters_result.fit(FootballPlayerRecords_Clustering)
  
  FootballPlayerRecords['Cluster_no'] = Clusters_result.labels_
  
#------------------------Step 10: Finding players similar to Song -------------------------------
  
  Song_cluster = FootballPlayerRecords.loc[FootballPlayerRecords['First_Name']=='Song','Cluster_no']
  
  Players_similar_to_song =FootballPlayerRecords.loc[FootballPlayerRecords['Cluster_no']==0,]
                                                           
  
  Players_similar_to_song
  
  
  
#------------------------------Different Approch:Using Hierarchical cluster --------------
  
  # Using the dendrogram to find the optimal number of clusters
   dendrogram = sch.dendrogram(sch.linkage(FootballPlayerRecords_Clustering, method = 'ward'))
   plt.title('Dendrogram')
   plt.xlabel('Customers')
   plt.ylabel('Euclidean distances')
   plt.show()
      
   # Fitting Hierarchical Clustering to the dataset
  
   hc = AgglomerativeClustering(n_clusters = 8, affinity = 'euclidean', linkage = 'ward')
   hc.fit(FootballPlayerRecords_Clustering)
   
   FootballPlayerRecords['Cluster_no'] = hc.labels_
   
   Song_cluster = FootballPlayerRecords.loc[FootballPlayerRecords['First_Name']=='Song','Cluster_no']
  
   Players_similar_to_song =FootballPlayerRecords.loc[FootballPlayerRecords['Cluster_no']==1,]
                                                           
  
   Players_similar_to_song
    